src/storage-types/src/connections/aws.rs

```diff
@@ -136,58 +227,119 @@ impl AwsConfig {
+                    .ok_or_else(|| anyhow!("missing secret key"))?;
+
+                let credentials = aws_credential_types::Credentials::from_keys(
+                    access_key,
+                    secret_key,
+                    temp_credentials.session_token().map(|t| t.to_string()),
+                );
+
+                let session_name = session_name.to_owned().unwrap_or("materialize".to_string());
+                let mut role = AssumeRoleProvider::builder(arn).session_name(session_name);
+
+                if let Some(region) = &region {
+                    role = role.region(region.clone());
+                }
+
+                if let Some(external_id_suffix) = external_id_suffix {
+                    role = role.external_id(mz_principal.get_external_id(external_id_suffix));
+                }
+
+                SharedCredentialsProvider::new(role.build(credentials))
             }
-            cred_provider = SharedCredentialsProvider::new(role.build(cred_provider));
-        }
+        };
 
         let mut loader = aws_config::from_env()
             .region(region)
             .credentials_provider(cred_provider);
         if let Some(endpoint) = &self.endpoint {
             loader = loader.endpoint_url(endpoint);
         }
-        loader.load().await
+        Ok(loader.load().await)
     }
 
-    #[allow(clippy::unused_async)]
     pub(crate) async fn validate(
         &self,
-        _id: GlobalId,
-        _connection_context: &ConnectionContext,
+        id: GlobalId,
+        connection_context: &ConnectionContext,
     ) -> Result<(), anyhow::Error> {
-        Err(anyhow!("Validating SSH connections is not supported yet"))
+        match &self.auth {
+            AwsAuth::Credentials(_) => {
+                let aws_config = self
+                    .get_sdk_config(connection_context.secrets_reader.as_ref(), None)
+                    .await?;
+
+                let sts_client = aws_sdk_sts::Client::new(&aws_config);
+                let _ = sts_client.get_caller_identity().send().await?;
+                Ok(())
+            }
+            AwsAuth::AssumeRole(_) => {
+                let aws_config = self
+                    .get_sdk_config(
+                        connection_context.secrets_reader.as_ref(),
+                        Some(id.to_string()),
+                    )
+                    .await?;
+
+                let sts_client = aws_sdk_sts::Client::new(&aws_config);
+                let _ = sts_client.get_caller_identity().send().await?;
+
+                let aws_config_without_external_id = self
+                    .get_sdk_config(connection_context.secrets_reader.as_ref(), None)
+                    .await?;
+
+                let sts_client = aws_sdk_sts::Client::new(&aws_config_without_external_id);
+                if sts_client.get_caller_identity().send().await.is_ok() {
+                    Err(anyhow!("Validate succeeded without external_id!"))
```

Comment:
This will be a user facing error and so should follow the error message style guide: https://github.com/MaterializeInc/materialize/blob/main/doc/developer/style.md#error-message-style. In particular a hint here would be really useful, since I imagine this will be quite a confusing error message for most users.

Reply:
Actually I think I will add a new error type for SDK errors.

Reply:
The validate method currently returns `Result<(), anyhow::Error>`, which eventually gets mapped to `AdapterError::Unstructured` with `SqlState::INTERNAL_ERROR` and no way to pass hint or detail for the error response. I will take a stab at refactoring that in a separate PR with a structured Error type so that hint and details can be passed properly. 

In this PR I am leaving things to be `anyhow!()`.